### ISSUE TITLE
GI Bill CT: Google Analytics not firing correctly

### ISSUE THREAD

**Issue by saintsoup52** [department-of-veterans-affairs/vets-website#5217]
In the WIP version of the tool it appears that Google Analytics is no longer firing for "EO Principles of Excellence" account. The account number is UA-46533988-2.

Additionally a number of the events seem to be disabled. For example:
* What "learn more"s are folks clicking on
* What schools are users selecting (by name versus facility code)
* What search terms are users using to find schools 

This is how VA tracks historical data on GI Bill CT usage and the way we can have great articles like these:
http://www.blogs.va.gov/VAntage/35078/gi-bill-comparison-tool-breaking-records-helping-veterans/

@mphprogrammer @saneshark

**Comment by saneshark** [comment]
Need to have conversation with stakeholders on this one and any reports that they are currently doing. We need to get a sense of how big this issue is. How long will it take to comb through and identify how the reporting and data analytics are affected.

@koconnor719 do you know who the specific stakeholders would be on this so that we can identify the reporting that they are doing and how we might be breaking their analytics.

**Comment by saneshark** [comment]
@eli-lewis I think this is one we need to have a stakeholder conversation around as soon as possible to get a good grasp on what specifically might change or break based on the design changes that will go into affect once we launch. I'm tagging you here for awareness, because sometime early next week would ideally be coordinating with the appropriate stakeholders to get this issue ready for actionable improvements by the front-end team to ensure we're addressing the points @saintsoup52 identified and any he may have missed.

**Comment by ayaleloehr** [comment]
Also tagging @robertfairhead as he has done most of the google analytics implementation across vets.gov and might be able to help get things in place.

**Comment by mphprogrammer** [comment]
Actually, I'd really be interested in assisting Robert with this, if there is anything a noob can do ...

**Comment by koconnor719** [comment]
@saneshark responding on slack

**Comment by koconnor719** [comment]
@saintsoup52 patrick who in VA EDU was using this report in the past?  Eddie and Jude? did it work at one time and now it doesn't?  (sorry i'm coming in late--if these questions don't make sense, please set me straight:)

**Comment by saintsoup52** [comment]
The analytics work in the production version but not in the WIP version.  Specifically, the POE google analytics account was left off for some reason... the DAP and vets.gov analytics work but not the POE ones. Additionally, a number of "events" will be broken in the new version.

As for who uses the reports... Stephanie and her team use the POE analytics for cabinet level reports to the WH among other reports. The main users were Eddie and Donald... however this is now John McNeal (who has been the main POC for a couple of weeks now).

**Comment by saintsoup52** [comment]
Here are some of the event categories we are currently tracking:
1. **Form change submits:** What eligibility and calculator questions are users interacting with.
  * gi-bill-chapter interactions
  * cumulative-service interactions
  * military-status interactions

2. **Type Ahead Clicks:** Are users clicking on the type ahead of the search results and what are the schools names being selected.
  * ARIZONA STATE UNIVERSITY-TEMPE
  * SAN DIEGO STATE UNIVERSITY
  * HARVARD UNIVERSITY

3. **Outbound links:** This currently works in both versions and is standard to VA websites.
  * OutBound to: See YR Rates
  * OutBound to: Student Complaints
  * OutBound to: Apply for GI Bill

4. **Info Icons:** Which info icons are users clicking on the most
  * eligible-for-kicker-info
  * tuition-fees-info
  * yellow-ribbon-info

5. **Click through to Profile:** What path does a user take to get to the school profile page
  * User Click on Learn More Button
  * User Click on School Name

![image](https://cloud.githubusercontent.com/assets/4397508/24600703/c9e8867a-1823-11e7-8657-301b07d85613.png)

**Comment by koconnor719** [comment]
Thanks😀



Sent with Good (www.good.com)


-----Original Message-----

**Comment by robertfairhead** [comment]
Sorry, I was sick last week so just catching up now.

We can re-use our Tag Manager instance to send the data to their account. It can be configured to send events to any GA account. That's preferable to inputing a third analytics snippet since GA uses some globally scoped variables by default, which makes setup a pain and your implementation brittle.

Can @mphprogrammer, @saneshark or someone point me at the new front-end code and let me know if there's an instance on dev/staging I can test against? I can take a quick breeze through and have a better idea of what the setup would be like. Sending custom events is generally pretty easy (normally 1 line of code per event), but we just have to be sure that there's an appropriate hook (e.g., React Action or JS event [onClick, etc]) that we can tap into to do it. Assuming there are, this is a couple hours of work to implement and test.

Thanks!

**Comment by koconnor719** [comment]
@saneshark the VA stakeholder to work w/ on this are Stephanie price and John McNeal.

**Comment by U-DON** [comment]
@robertfairhead The front-end code is in vets-website/src/js/gi. On staging, it's at https://staging.vets.gov/gi-bill-comparison-tool.

**Comment by robertfairhead** [comment]
@U-DON thanks. i'll look at it on wednesday and get you some better thoughts on this then.

**Comment by saneshark** [comment]
@robertfairhead Any updates here? This one is going to a blocker for launch. Based on your findings we might need to push that date out, we might need to push our launch window out.

**Comment by robertfairhead** [comment]
@saneshark working on it now

**Comment by robertfairhead** [comment]
Initial thoughts @saneshark :

Basic page tracking is simple. I can add a configuration item to send all the page view, user, etc. to UA-46533988-2 that'll take 10 minutes.

For the 5 custom events:

1. Form change submits: Looks like those three boxes are in EligibiltyForm.jsx and all have onclick handlers that go to eligibilityChange action. In actions/index.js we could update that function to emit the field and value when a new value is selected. That's quick, maybe 3 lines of code.

2. Type Ahead Clicks: Looks like that's KeywordSearch.jsx and it's got handleSuggestionSelected(event, data) that looks to be called when they click a suggested value that we could add in a couple lines to fire off the event. But @U-DON or someone would need to verify that's the right function. I'm not familiar enough with the react autosuggest component to be 100% on it.

3. Outbound links i can config in tag manager so no code there. maybe 20 minutes effort for me.

4. Info Icons: I couldn't tell what those were in the new UI. Can some drop me a screenshot or something so I can know for sure?

5. For this, i only see the click on title in the new UI. I didn't see a "Learn more button" We'll see openings of the profile pages as page views so that'll get handled automatically with the basic setup. Let me know if I'm missing something there.

My guess on LOE: half hour of config work for me (I'll start now). Plus maybe 6 lines of code that'll need to be reviewed. I can do a quick PR or pair with someone to do it (maybe a hour to write/review). If there's somebody who can do the review for the front-end stuff, I can make the time this afternoon to knock that out.

Plus someone should verify that 4 & 5 no longer apply in the new UI.

**Comment by robertfairhead** [comment]
@saintsoup52 I have the config changes ready to send you the basic tracking info for GIBCT pages. Our site sends traffic from both production and non-production environments (to allow for testing) and then our GA properties filter out the non-production traffic. I assume that you handled filtering out the non-production traffic already (since that's how the prior GIBCT tool worked) but wanted to mention it before we started pushing the new traffic to prevent any confusion down the road.

**Comment by saintsoup52** [comment]
@robertfairhead Thank you for the heads up. I will keep an eye out for the traffic.

As for the the specific events listed above... the fifth one concerning the "learn more" button doesn't exist in the new design. Hope that helps.

**Comment by melwoodard** [comment]
@robertfairhead sorry for the delay - #4 refers to what are now "learn more" text links that appear next to certain field labels and data values throughout the tool. These have replaced info icons (see the tool in production for comparison)

**Comment by robertfairhead** [comment]
@melwoodard any chance you have some examples of them? URLs to staging would be great but even a screen shot is useful. If not, I can try to find them but having a starting point is super helpful.

**Comment by saintsoup52** [comment]
Here you go... Just select any school and look in the calculator section.

![image](https://cloud.githubusercontent.com/assets/4397508/24802146/c6810134-1b74-11e7-80c3-0b9804c993ea.png)

**Comment by robertfairhead** [comment]
Got it. Very helpful. Those look like they all filter back to the showModal action. We should be able to catch them all there. @saneshark @U-DON i'll add that in to the branch i sent you yesterday (https://github.com/department-of-veterans-affairs/vets-website/tree/gibct-analytics)

**Comment by robertfairhead** [comment]
Ok status check on this:

- [X] GA Configured to send basic data to UA-46533988-2
- 1. Form Changes
  - [X] GA configured to send events to Vets.gov and EO Principles of Excellence accounts
  - [X] Initial code on gibct-analytics branch
  - [ ] Code reviewed and merged
- 2. Type Ahead Clicks
  - [X] GA configured to send events to Vets.gov and EO Principles of Excellence accounts
  - [X] Initial code on gibct-analytics branch
  - [ ] Code reviewed and merged
- 3. Outbound links
  - [X] GA configured to send events to Vets.gov and EO Principles of Excellence accounts
- 4. Learn More Links (formerly Info Icons)
  - [X] GA configured to send events to Vets.gov and EO Principles of Excellence accounts
  - [X] Initial code on gibct-analytics branch
  - [ ] Code reviewed and merged
- 5. Click through to Profile
  - [X] Confirmed this no longer applies

So primary next steps are for Kam and Eugene to review the branch to be sure the event collection is really in the right place then merge. Once on dev/staging I can test to be sure they're firing correctly before it hits production.

In addition, via Slack, Kam noted the following:

- Collecting search parameters: These will show up as query params (e.g., /gi-bill-comparison-tool/search?name=indianapolis) in the URLs passed to the EO Principles of Excellence account

- Collecting school names: The profile page views will pass them as the Page Title attributes, which will be passed to the the EO Principles of Excellence account

**Comment by leannammiller** [comment]
who will be responsible for reviewing and merging code?

**Comment by saneshark** [comment]
Woohoo! sorry for commenting on a closed thread but this one was one of the ones that had me most concerned! Thanks @U-DON and @robertfairhead

**Comment by saintsoup52** [comment]
Can we please have a discussion about this? 

I was able to test the analytics.  They are firing for POE analytics account... thank you.

Unfortunately, many of the events are not working as expected.

**1) Outbound link events**
In the current version of the tool there are unique labels to the outbound links. Those labels are not going to the POE account. For example "OutBound to: Apply for GI Bill." Oddly enough those labels are working correctly with the VA analytics account... so hopefully this should be an easy change.

**2) Learn mores events**
Not all the learn mores are firing analytics or maybe they are firing as "undefined".

![image](https://cloud.githubusercontent.com/assets/4397508/25285783/6b93efc8-2689-11e7-9126-85784dd495f3.png)

@leannammiller @saneshark @mphprogrammer

**Comment by leannammiller** [comment]
@robertfairhead do you mind helping us with this?

**Comment by robertfairhead** [comment]
sure, i can take a look on monday. 

1. sounds like its probably a config issue with the event labels. From what Patrick posted it sounds like they were sending the link text over, which should be a straightforward config option.

2. I'll have to take a look at this one. It's hooked into underlying learn more machinery so it should get them all. @saintsoup52 if you have specific examples of which learn mores aren't firing correctly that would help debug.

**Comment by melwoodard** [comment]
@robertfairhead did you figure out what the issue is here?

**Comment by robertfairhead** [comment]
The event labels on 1 I tweaked. I didn't deploy the fix for that yet because I'm still a little stumped on 2. They seem to be firing ok, so I was going to play with it a bit this morning to see if I can figure out which ones aren't working for them. If I can't, I'll at least post the update that should address 1.

**Comment by robertfairhead** [comment]
Ok, I deployed the config fix for the outbound links. They should be identically formatted to the prior version.

https://github.com/department-of-veterans-affairs/vets-website/pull/5421 is the PR with the fix for the "Learn More" issue. All of the "Learn More" modal were correctly firing (that I could find) but the event was erroneously also getting fired when the modal was closed. Those were the ones showing up as "undefined" and creating the appearance of missed events (from what I can see). The PR will prevent those false events from getting fired.
